llvm/lib/Target/SPIRV/SPIRVTargetMachine.cpp

```diff
@@ -178,6 +266,9 @@ void SPIRVPassConfig::addIRPasses() {
     addPass(createSPIRVStructurizerPass());
   }
 
+  if (TM.getOptLevel() > CodeGenOptLevel::None)
+    addPass(createInferAddressSpacesPass(AddressSpace::Generic));
```

Comment:
Not sure why this is a pass parameter to InferAddressSpaces, and a TTI hook

Reply:
Because if one invokes the pass directly via opt there's no way but the TTI query to set Flat/Generic to anything but 0, and because making it explicit at the point of construction rather than relying on that seems somewhat more self documenting.

Reply:
Out of curiosity, why do invoke this pass twice: in the middle-end and code-gen?

Reply:
I remember we had some phase ordering issues where we needed to run this multiple times. I'm not sure what the current status is. We certainly need to run this after inlining